### PR TITLE
procedural point instancer

### PR DESCRIPTION
**Pull request by kreppene**
**Changes proposed in this pull request**
A proposed implementation of the pointInstancer, currently velcoty and primvars is not supported in this pr

**Additional context**
Screenshot_3

### CODE REVIEW

**Comment by sebastienblor** [comment]
Hi Raymond, thanks for the PR !
One question, in which use case do you need to create a new `usd` node in the scene pointing at this object_path ? You seem to be doing it if the corresponding object doesn't exist in the arnold scene but I wanted to know if this was a valid use case. I'm afraid this will be a problem, because the translator can be called without any procedural parent, so duplicating the procedural can't really work in the generic case. 

In order to ensure that the proto node is already converted to arnold, you can take a look at the code in `UsdArnoldReadGeometryLight`, where we need to ensure that the mesh is created in Arnold before the geometry light points at it. I see a comment saying that the data is exported twice but the comment is actually deprecated and should be removed.

**Comment by sebastienblor** [comment]
All the explicit calls to 
`AiNode("nodeType")` 
should be replaced by 
`getNodeToConvert(reader, "nodeType", "nodeName", create, convert);`

as the created nodes need to be registered, and must eventually be given the parent procedural.

File: translator/reader/read_geometry.cpp [hunk 1]
```diff
@@ -206,6 +211,84 @@ void UsdArnoldReadCube::read(const UsdPrim &prim, UsdArnoldReader &reader, bool
...
+        const double* matrixArray = xforms[index].GetArray();
+        xform.insert(xform.end(), matrixArray, matrixArray + 16);
+        //TODO caluclate velcoty when avalbile
+
+
+        // instance the prototype
+        int id = *iter;
+
+        AtNode *arnold_instance = AiNode("ginstance");
+        std::string instance_name = std::string(AiNodeGetName(protoNodes[id])) + std::string("_") + std::to_string(index);
```
[the remark this answers is not in the document]
Comment (sebastienblor) [reply translator/reader/read_geometry.cpp]: right, it should rather be the name of the PointInstancer node, plus the index of the instance

**Comment by kreppene** [comment]
Hi, 
Great feedback, i try to get trough them all shortly,

> Hi Raymond, thanks for the PR !
> One question, in which use case do you need to create a new usd node in the scene pointing at this object_path ? You seem to be doing it if the corresponding object doesn't exist in the arnold scene but I wanted to know if this was a valid use case.

I was expecting this question, 
theirs probably a better way to do this,  but i could not find out how to go about instance a hierarchy of objects, like the example below, if you where to instance the tree with all its children,
Easiest way i found was just to create a new usd procedual containing the tree with all the branches, and then instance that tree around
i think this is what they are doing also inn the OpenWalter arnold procedural.
Maby thers a better way to do this?

>  I'm afraid this will be a problem, because the translator can be called without any procedural parent, so duplicating the procedural can't really work in the generic case.
i'm not sure i understand this part, could you elaborate on this?

|--tree
          |---branch_01
                    |---smaller_branches_01

**Comment by sebastienblor** [comment]
> I was expecting this question,
> theirs probably a better way to do this, but i could not find out how to go about instance a hierarchy of objects, like the example below, if you where to instance the tree with all its children,
> Easiest way i found was just to create a new usd procedual containing the tree with all the branches, and then instance that tree around
> i think this is what they are doing also inn the OpenWalter arnold procedural.
> Maby thers a better way to do this?

It's indeed a good solution to address not only this use case, but also the case of recursive point instancers. So the only thing to change is to avoid looking for the procedural parent and instead store the filename that is called in `read()` as I said in another comment. The "translator" module was separated from the procedural because it can be invoked in other contexts (just converting a scene without any procedural involved).

**Comment by sebastienblor** [comment]
Hi Raymond, 
thanks a lot for the changes. There are still a few things to adjust, do you mind if we push other changes to your fork branch before we merge it ?
Thanks

**Comment by kreppene** [comment]
Be my guest Sebastian

**Comment by sebastienblor** [comment]
Did some changes in the last commit : 

- moved the UsdPointInstancerReader function at the end of the file, as it was in the middle of a block of implicit shapes translators
- fixed the indentation
- made the nodes creation more robust for the different scenarios in which it can be called. I said above that we should use `getNodeToConvert`  but after a second thought it makes more sense to call directly `reader.createArnoldNode`  as we don't need the tests that are done in the first function.
- set the filename in `UsdArnoldReader::read` instead of setting it from the procedural code, as we need this to work even without a procedural. Made this filename private and added a get function.
- consider the time range given by the procedural for motion blur
- take into account the invisibleIds  as suggested by @sirpalee 
- added more comments as the creation of another usd procedural is not obvious to understand
- added test_0022 with a point instancer node

Note that the use case of another child usd procedural being created doesn't actually work, because of another issue I just spotted while debugging. I'll create a ticket for that. 

The problem is that when the created procedural will check if the nodes already exist, then `AiNodeLookUpByName` will return the node that was created by the original procedural, given that it's its parent, and that name lookups check recursively in each parent scope. Also, once we fix that issue, it will still mean that we're creating the proto nodes twice (once by the original procedural, twice by the child one with object_path). So we still need to give this solution some love for it to work properly

**Comment by sebastienblor** [comment]
Note that #189 would be needed to avoid the issues I mentioned, when the proto is a xform node and we need to create another `usd` proc

**Comment by sebastienblor** [comment]
I merged master and fixed the conflicts by porting the latest changes.
I see things are being converted in the procedural as expected, but when I run test_0023 arnold freezes during rendering. Still need to debug this

**Comment by sebastienblor** [comment]
After some debugging, we found that the hangs in test 23 were coming from the core, in a ticket that is being addressed. Until then, to workaround the issue, we need to ensure the parallel init is enabled in this procedural, which is implemented in #203 .  This solves the hangs and allows the point instancer to be translated properly

**Comment by sebastienblor** [comment]
I merged master in order to get the fix from #203 and also added `test_0024` that is identical to `test_0023` except that the prototype node is a Xform and thus requires to create a child usd procedural

**Comment by sebastienblor** [comment]
After discussing with Pal, I've made it in last commit so that the overrides array is also copied to the child usd procedurals

**Pull request merged**